lib/src/cancelable_operation.dart

```diff
@@ -39,6 +39,59 @@ class CancelableOperation<T> {
     return completer.operation;
   }
 
+  /// Creates a [CancelableOperation] wrapping [subscription].
+  ///
+  /// This overrides [subscription.onDone] and [subscription.onError] so that
+  /// the returned operation will complete when the subscription completes or
+  /// emits an error. When this operation is canceled or when it emits an error,
+  /// the subscription will be canceled (unlike
+  /// `CancelableOperation.fromFuture(subscription.asFuture())`).
+  static CancelableOperation<void> fromSubscription(
+      StreamSubscription<void> subscription) {
+    var completer = CancelableCompleter<void>(onCancel: subscription.cancel);
+    subscription.onDone(completer.complete);
+    subscription.onError((Object error, StackTrace stackTrace) {
+      subscription.cancel();
+      completer.completeError(error, stackTrace);
+    });
+    return completer.operation;
+  }
+
+  /// Returns a [CancelableOperation] that completes with the value of the first
+  /// of [operations] to complete.
+  ///
+  /// Once any of [operations] completes, its result is forwarded to the
+  /// returned [CancelableOperation] and the rest are cancelled. When the
+  /// returned operation is cancelled, all the [operations] are cancelled as
+  /// well.
+  static CancelableOperation<T> race<T>(
+      Iterable<CancelableOperation<T>> operations) {
+    operations = operations.toList();
+    if (operations.isEmpty) throw ArgumentError("operations can't be empty");
```

Comment:
I'd prefer using the `value` constructor:
```dart
ArgumentError.value(operations, "operations", "Must not be empty");
```
I can see that the "safe toString" of the `operations` iterable won't be very useful in the error's `toString`. At least consider using
```dart
  throw ArgumentError("Must not be empty", "operations");
```
to include the variable name. (And capitalize the message, it'll occur after a colon. Use "must not" for requirements, not "can't be", because obviously it *could* be empty since we got here.)